File: framework/languages/node/samples/SupportChat.Ts/Shared/Contracts/messages.ts
type SupportRole = 'Agent' | 'Customer';
type ConversationStatus = 'WaitingForAgent' | 'Active' | 'WaitingForClose' | 'Closed';

class AuthenticateReq { constructor(readonly accessToken: string) {} }
type AuthenticateRes = { actorId: string; displayName: string; role: SupportRole };
class AuthenticateUserReq { constructor(readonly accessToken: string) {} }
type AuthenticateUserRes = {
  accepted: boolean;
  actorId?: string;
  displayName?: string;
  role?: SupportRole;
  reason?: string;
};
class OpenConversationApiReq {
  constructor(readonly customerActorId: string, readonly customerDisplayName: string, readonly subject: string) {}
}
type OpenConversationApiRes = { conversationId: string; status: ConversationStatus };
class AllocateConversationReq {
  constructor(readonly customerActorId: string, readonly customerDisplayName: string, readonly subject: string) {}
}
type AllocateConversationRes = { conversationId: string; status: ConversationStatus };
class AssignAgentReq { constructor(readonly conversationId: string, readonly agentActorId: string) {} }
type AssignAgentRes = { conversationId: string; assigned: boolean };
class OpenConversationReq { constructor(readonly subject: string) {} }
type OpenConversationRes = { conversationId: string; state: ConversationState };
class SetAgentAvailableReq { constructor(readonly isAvailable: boolean) {} }
type SetAgentAvailableRes = { isAvailable: boolean };
class JoinConversationReq { constructor(readonly conversationId: string) {} }
type JoinConversationRes = { state: ConversationState };
class SendChatMessageReq { constructor(readonly conversationId: string, readonly text: string) {} }
type SendChatMessageRes = { message: ChatMessage; state: ConversationState };
class SetTypingReq { constructor(readonly conversationId: string, readonly isTyping: boolean) {} }
class CloseConversationReq { constructor(readonly conversationId: string, readonly reason?: string) {} }
type CloseConversationRes = { state: ConversationState };
type ParticipantJoinedNotify = { conversationId: string; actorId: string; role: SupportRole; state: ConversationState };
type ConversationAssignedNotify = { conversationId: string; state: ConversationState };
type ChatMessageNotify = { conversationId: string; message: ChatMessage; state: ConversationState };
type TypingChangedNotify = { conversationId: string; actorId: string; isTyping: boolean; state: ConversationState };
type ConversationIdleNotify = { conversationId: string; state: ConversationState };
type ConversationClosedNotify = { conversationId: string; state: ConversationState };
type ConversationState = {
  conversationId: string;
  subject: string;
  status: ConversationStatus;
  customerActorId: string;
  agentActorId?: string;
  lastMessageSeq: number;
  lastMessageAtUnixMs?: number;
  idleDeadlineUnixMs?: number;
};
type ChatMessage = {
  conversationId: string;
  messageSeq: number;
  senderActorId: string;
  text: string;
  sentAtUnixMs: number;
};

const SupportChatRoles = {
  Agent: 'Agent',
  Customer: 'Customer'
} as const;

const ConversationStatuses = {
  WaitingForAgent: 'WaitingForAgent',
  Active: 'Active',
  WaitingForClose: 'WaitingForClose',
  Closed: 'Closed'
} as const;

const PacketNames = {
  authenticateReq: 'AuthenticateReq',
  authenticateRes: 'AuthenticateRes',
  authenticateUserReq: 'AuthenticateUserReq',
  authenticateUserRes: 'AuthenticateUserRes',
  openConversationApiReq: 'OpenConversationApiReq',
  openConversationApiRes: 'OpenConversationApiRes',
  allocateConversationReq: 'AllocateConversationReq',
  allocateConversationRes: 'AllocateConversationRes',
  assignAgentReq: 'AssignAgentReq',
  assignAgentRes: 'AssignAgentRes',
  openConversationReq: 'OpenConversationReq',
  openConversationRes: 'OpenConversationRes',
  setAgentAvailableReq: 'SetAgentAvailableReq',
  setAgentAvailableRes: 'SetAgentAvailableRes',
  joinConversationReq: 'JoinConversationReq',
  joinConversationRes: 'JoinConversationRes',
  sendChatMessageReq: 'SendChatMessageReq',
  sendChatMessageRes: 'SendChatMessageRes',
  setTypingReq: 'SetTypingReq',
  closeConversationReq: 'CloseConversationReq',
  closeConversationRes: 'CloseConversationRes',
  participantJoinedNotify: 'ParticipantJoinedNotify',
  conversationAssignedNotify: 'ConversationAssignedNotify',
  chatMessageNotify: 'ChatMessageNotify',
  typingChangedNotify: 'TypingChangedNotify',
  conversationIdleNotify: 'ConversationIdleNotify',
  conversationClosedNotify: 'ConversationClosedNotify'
} as const;

function authenticate(accessToken: string): AuthenticateReq {
  return new AuthenticateReq(accessToken);
}

function authenticateUser(accessToken: string): AuthenticateUserReq {
  return new AuthenticateUserReq(accessToken);
}

function openConversation(subject: string): OpenConversationReq {
  return new OpenConversationReq(subject);
}

function allocateConversation(customerActorId: string, customerDisplayName: string, subject: string): AllocateConversationReq {
  return new AllocateConversationReq(customerActorId, customerDisplayName, subject);
}

function assignAgent(conversationId: string, agentActorId: string): AssignAgentReq {
  return new AssignAgentReq(conversationId, agentActorId);
}

function setAgentAvailable(isAvailable: boolean): SetAgentAvailableReq {
  return new SetAgentAvailableReq(isAvailable);
}

function joinConversation(conversationId: string): JoinConversationReq {
  return new JoinConversationReq(conversationId);
}

function sendChatMessage(conversationId: string, text: string): SendChatMessageReq {
  return new SendChatMessageReq(conversationId, text);
}

function setTyping(conversationId: string, isTyping: boolean): SetTypingReq {
  return new SetTypingReq(conversationId, isTyping);
}

function closeConversation(conversationId: string, reason?: string): CloseConversationReq {
  return new CloseConversationReq(conversationId, reason);
}

export {
  PacketNames,
  AuthenticateReq,
  AuthenticateUserReq,
  OpenConversationApiReq,
  AllocateConversationReq,
  AssignAgentReq,
  OpenConversationReq,
  SetAgentAvailableReq,
  JoinConversationReq,
  SendChatMessageReq,
  SetTypingReq,
  CloseConversationReq,
  SupportChatRoles,
  ConversationStatuses,
  authenticate,
  authenticateUser,
  openConversation,
  allocateConversation,
  assignAgent,
  setAgentAvailable,
  joinConversation,
  sendChatMessage,
  setTyping,
  closeConversation
};

export type {
  SupportRole,
  ConversationStatus,
  AuthenticateRes,
  AuthenticateUserRes,
  OpenConversationApiRes,
  AllocateConversationRes,
  AssignAgentRes,
  OpenConversationRes,
  SetAgentAvailableRes,
  JoinConversationRes,
  SendChatMessageRes,
  CloseConversationRes,
  ParticipantJoinedNotify,
  ConversationAssignedNotify,
  ChatMessageNotify,
  TypingChangedNotify,
  ConversationIdleNotify,
  ConversationClosedNotify,
  ConversationState,
  ChatMessage
};

File: framework/languages/dotnet/samples/SupportChat/Shared/Contracts/Messages.cs
namespace SupportChat.Shared.Contracts;

public sealed record AuthenticateReq(string AccessToken);

public sealed record AuthenticateRes(
    string ActorId,
    string DisplayName,
    string Role);

public sealed record AuthenticateUserReq(string AccessToken);

public sealed record AuthenticateUserRes(
    bool Accepted,
    string? ActorId,
    string? DisplayName,
    string? Role,
    string? Reason);

public sealed record EnsureSupportUserActorReq(
    string ActorId,
    string DisplayName,
    string Role,
    string ParticipantId);

public sealed record ActorRefSnapshot(
    byte[] NodeRid,
    string ActorId,
    ulong Generation);

public sealed record EnsureSupportUserActorRes(ActorRefSnapshot Actor);

// One agent serves many conversations, so each assigned conversation gets its own
// per-conversation agent actor (ParticipantId = the agent's roster id). The session
// asks the Support server to create that actor and join it into the ConversationSpot.
public sealed record EnsureAgentConversationReq(
    string RosterActorId,
    string DisplayName,
    string ConversationId);

public sealed record EnsureAgentConversationRes(
    ActorRefSnapshot Actor,
    ConversationState State);

public sealed record OpenConversationApiReq(
    string CustomerActorId,
    string CustomerDisplayName,
    string Subject);

public sealed record OpenConversationApiRes(
    string ConversationId,
    string Status);

public sealed record AllocateConversationReq(
    string CustomerActorId,
    string CustomerDisplayName,
    string Subject);

public sealed record AllocateConversationRes(
    string ConversationId,
    string Status);

public sealed record OpenConversationReq(string Subject);

public sealed record OpenConversationRes(
    string ConversationId,
    ConversationState State);

public sealed record SetAgentAvailableReq(bool IsAvailable);

public sealed record SetAgentAvailableRes(bool IsAvailable);

// ConversationId travels as stream message metadata (§9.2), not in these bodies.
public sealed record JoinConversationReq();

public sealed record JoinConversationRes(ConversationState State);

public sealed record SendChatMessageReq(string Text);

public sealed record SendChatMessageRes(
    ChatMessage Message,
    ConversationState State);

// SetTyping is a one-way fire-and-forget send: no response record.
public sealed record SetTypingReq(bool IsTyping);

public sealed record CloseConversationReq(string? Reason);

public sealed record CloseConversationRes(ConversationState State);

public sealed record ParticipantJoinedNotify(
    string ConversationId,
    string ActorId,
    string Role,
    ConversationState State);

public sealed record ConversationAssignedNotify(
    string ConversationId,
    ConversationState State);

public sealed record ChatMessageNotify(
    string ConversationId,
    ChatMessage Message,
    ConversationState State);

public sealed record TypingChangedNotify(
    string ConversationId,
    string ActorId,
    bool IsTyping,
    ConversationState State);

public sealed record ConversationIdleNotify(
    string ConversationId,
    ConversationState State);

public sealed record ConversationClosedNotify(
    string ConversationId,
    ConversationState State);

public sealed record ConversationState(
    string ConversationId,
    string Subject,
    string Status,
    string CustomerActorId,
    string? AgentActorId,
    ulong LastMessageSeq,
    long? LastMessageAtUnixMs,
    long? IdleDeadlineUnixMs);

public sealed record ChatMessage(
    string ConversationId,
    ulong MessageSeq,
    string SenderActorId,
    string Text,
    long SentAtUnixMs);